### PR TITLE
feat(util): atomWithStorage

### PR DESCRIPTION
**Pull request by sandren**
There are examples of how to add persistence to individual atoms in the docs, but I would like to create a robust `atomWithStorage` utility that covers most use cases, including `localStorage` or `sessionStorage` for React as well as `AsyncStorage` for React Native. Also note that the stored value is intentionally populated in the `onMount` method rather than the base `atom()` initialization to support `atomWithStorage` usage in SSR contexts such as Next.js and Gatsby. The first commit is a work in progress.

**Progress**
* [x]  impl
* [x]  tests
* docs -> a separate pr

### CODE REVIEW

File: src/utils/atomWithStorage.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { atom, WritableAtom } from 'jotai'
+import type { SetStateAction } from '../core/types'
+
+export function atomWithStorage<Value>(key: string, initialValue: Value) {
+  type Update = SetStateAction<Value>
```
Comment (dai-shi) [inline src/utils/atomWithStorage.ts:5]: Do we accept function update? If so, L24-L26 should be implemented so. There's no magic.

**Comment by dai-shi** [comment]
The WIP version looks actually good, and is worth a guide in docs.

To make if a flexible util, we want to abstract the storage api.
```ts
type Storage<Value> = {
  getItem: (key: string) => Value | Promise<Value>
  setItem: (key: string, newValue: Value) => void | Promise<void>
}

export function atomWithStorage<Value>(key: string, default: Value, getStorage?: () => Storage<Value>): PrimitiveAtom<Value>
```

File: src/utils/atomWithStorage.ts (hunk 1 above)
Comment (sandren) [reply src/utils/atomWithStorage.ts:5]: Can you elaborate?

Comment (dai-shi) [reply src/utils/atomWithStorage.ts:5]: It has to be:
```ts
    (get, set, update: Update) => {
      const newValue = typeof update === 'function' ? update(get(baseAtom)) : update
      set(baseAtom, newValue)
      window.localStorage.setItem(key, JSON.stringify(newValue))
    }
```

**Review by dai-shi — APPROVED** [review]

**Comment by ghmeec** [comment]
this is awesome, i haven't been able to test in RN environment, would there be a problem since getItem and setItem methods are async?

**Comment by dai-shi** [comment]
@ghmeec We hoped this would support AsyncStorage, but not thoroughly tested. Would you give it a try?

**Comment by thinkdj** [comment]
Often, we need to store sensitive data like a token or basic authentication info for persistence, and storing such data plainly to localStorage seems less secure. Provision for a custom encode/decode method would be nice to have.

**Comment by thinkdj** [comment]
Often, we need to store sensitive data like a token or basic authentication info for persistence, and storing such data plainly to localStorage seems less secure. Provision for a custom hook for encrypt/decrypt would be great to see

**Comment by dai-shi** [comment]
You should be able to customize `storage`.

**Comment by thinkdj** [comment]
> You should be able to customize `storage`.

Ah! Missed this. Thank you. 

Reference for others:  Customizing `storage`: [Docs](https://jotai.org/docs/utils/atom-with-storage#parameters) & [sample](https://github.com/pmndrs/jotai/discussions/1220).

**Pull request merged**